relay-metrics/src/bucket.rs

```diff
@@ -614,19 +747,84 @@ pub struct Bucket {
 }
 
 impl Bucket {
-    /// Parses a single metric bucket from the JSON protocol.
-    pub fn parse(slice: &[u8]) -> Result<Self, ParseBucketError> {
-        serde_json::from_slice(slice).map_err(ParseBucketError)
+    /// Parse statsd-compatible payload of format
+    /// ```text
+    /// [<ns>/]<name>[@<unit>]:<value>|<type>[|#<tags>]`
+    /// ```
+    fn parse_str(string: &str, timestamp: UnixTimestamp) -> Option<Self> {
+        let mut components = string.split('|');
+
+        let (mri_str, values_str) = components.next()?.split_once(':')?;
+        let ty = components.next().and_then(|s| s.parse().ok())?;
+
+        let mri = parse_mri(mri_str, ty)?;
+        let value = match ty {
+            MetricType::Counter => BucketValue::Counter(parse_counter(values_str)?),
+            MetricType::Distribution => BucketValue::Distribution(parse_distribution(values_str)?),
+            MetricType::Set => BucketValue::Set(parse_set(values_str)?),
+            MetricType::Gauge => BucketValue::Gauge(parse_gauge(values_str)?),
+        };
+
+        let mut bucket = Bucket {
+            timestamp,
+            width: 0,
+            name: mri.to_string(),
+            value,
+            tags: Default::default(),
+        };
+
+        for component in components {
+            if let Some('#') = component.chars().next() {
+                bucket.tags = parse_tags(component.get(1..)?)?;
+            }
+        }
+
+        Some(bucket)
     }
 
-    /// Parses a set of metric bucket from the JSON protocol.
-    pub fn parse_all(slice: &[u8]) -> Result<Vec<Bucket>, ParseBucketError> {
-        serde_json::from_slice(slice).map_err(ParseBucketError)
+    /// Parses a single metric aggregate from the raw protocol.
+    ///
+    /// See the [`Bucket`] for more information on the protocol.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use relay_metrics::{Bucket, UnixTimestamp};
+    ///
+    /// let bucket = Bucket::parse(b"response_time@millisecond:57|d", UnixTimestamp::now())
+    ///     .expect("metric should parse");
+    /// ```
+    pub fn parse(slice: &[u8], timestamp: UnixTimestamp) -> Result<Self, ParseMetricError> {
+        let string = std::str::from_utf8(slice).or(Err(ParseMetricError(())))?;
+        Self::parse_str(string, timestamp).ok_or(ParseMetricError(()))
     }
 
-    /// Serializes the given buckets to the JSON protocol.
-    pub fn serialize_all(buckets: &[Self]) -> Result<String, ParseBucketError> {
-        serde_json::to_string(&buckets).map_err(ParseBucketError)
+    /// Parses a set of metric aggregates from the raw protocol.
```

Comment:
nit: technically, this function does not parse the metrics (as that's done lazily in the iterator). Should we update the docstrings? (This was not added in this PR, but I thought it's worth pointing it out.)

Reply:
It's fine, it returns an iterator over parsed buckets. The signature and doc comment should make it clear that this is a lazy iterator.